### PR TITLE
Ros2 devel

### PR DESCRIPTION
**Pull request by cuixiongyi**
This is the repo that is open sourced on github under MIT license. It communicates with BNO055 using i2c.
I adopted it to ROS2. I used our convention of package manifest and CMakeList. But kept the original naming convention, ROS parameter and formatting, since we are thinking about contribute this code back to the original repo.

### CODE REVIEW

**Comment by ToffeeAlbina-TRI** [comment]
Since we'd like to push this upstream and have it maintained as a long running ROS2 branch, updating the README.md to reflect that this is ROS2 makes sense.

**Comment by ToffeeAlbina-TRI** [comment]
I would either remove imu.launch or make it a ROS2 launch file... probably the latter.

**Comment by ToffeeAlbina-TRI** [comment]
There are parameters, right?  We should add a default parameters file (like we have in Outdoor_muppet_base/config for example)

**Comment by cuixiongyi** [comment]
> ere are parameters, right? We

I have one, config/default_muppet_imu_bno055_params.yaml
Although I should probably get rid of the 'muppet'.

**Comment by ToffeeAlbina-TRI** [comment]
This is a ROS2 node, remove the imu.launch file - it is not how ROS2 launch works.

File: imu_bno055/launch/imu.launch.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Copyright 2018 Toyota Research Institute.  All rights reserved.
+#
+# IF WE RELEASE THIS CODE, WE MAY USE THE FOLLOWING BOILERPLATE:
+#
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
```
Comment (ToffeeAlbina-TRI) [inline imu_bno055/launch/imu.launch.py]: Hmmm, if we're trying to upstream this, then this boilerplate stuff will have to be  removed... probably not suggesting Apache if the original package is MIT

File: imu_bno055/src/bno055_i2c_activity.cpp [hunk 2]
```diff
@@ -158,29 +182,29 @@ bool BNO055I2CActivity::spinOnce() {
         return false;
     }
 
-    sensor_msgs::Imu msg_raw;
+    sensor_msgs::msg::Imu msg_raw;
     msg_raw.header.stamp = time;
     msg_raw.header.frame_id = param_frame_id;
-    msg_raw.header.seq = seq;
+//    msg_raw.header.seq = seq;
```
Comment (ToffeeAlbina-TRI) [inline imu_bno055/src/bno055_i2c_activity.cpp]: why commented out instead of removed?

File: imu_bno055/src/bno055_i2c_activity.cpp [hunk 3]
```diff
@@ -158,29 +182,29 @@ bool BNO055I2CActivity::spinOnce() {
...
     msg_raw.angular_velocity.x = (double)record.raw_angular_velocity_x / 900.0;
     msg_raw.angular_velocity.y = (double)record.raw_angular_velocity_y / 900.0;
     msg_raw.angular_velocity.z = (double)record.raw_angular_velocity_z / 900.0;
 
-    sensor_msgs::MagneticField msg_mag;
+    sensor_msgs::msg::MagneticField msg_mag;
     msg_mag.header.stamp = time;
     msg_mag.header.frame_id = param_frame_id;
-    msg_mag.header.seq = seq;
+//    msg_mag.header.seq = seq;
```
Comment (ToffeeAlbina-TRI) [inline imu_bno055/src/bno055_i2c_activity.cpp]: ditto

File: imu_bno055/src/bno055_i2c_activity.cpp [hunk 4]
```diff
@@ -158,29 +182,29 @@ bool BNO055I2CActivity::spinOnce() {
...
     msg_mag.magnetic_field.x = (double)record.raw_magnetic_field_x / 16.0;
     msg_mag.magnetic_field.y = (double)record.raw_magnetic_field_y / 16.0;
     msg_mag.magnetic_field.z = (double)record.raw_magnetic_field_z / 16.0;
 
-    sensor_msgs::Imu msg_data;
+    sensor_msgs::msg::Imu msg_data;
     msg_data.header.stamp = time;
     msg_data.header.frame_id = param_frame_id;
-    msg_data.header.seq = seq;
+//    msg_data.header.seq = seq;
```
Comment (ToffeeAlbina-TRI) [inline imu_bno055/src/bno055_i2c_activity.cpp]: ditto

File: imu_bno055/src/bno055_i2c_activity.cpp [hunk 5]
```diff
@@ -199,16 +223,16 @@ bool BNO055I2CActivity::spinOnce() {
     msg_data.angular_velocity.y = (double)record.raw_angular_velocity_y / 900.0;
     msg_data.angular_velocity.z = (double)record.raw_angular_velocity_z / 900.0;
 
-    sensor_msgs::Temperature msg_temp;
+    sensor_msgs::msg::Temperature msg_temp;
     msg_temp.header.stamp = time;
     msg_temp.header.frame_id = param_frame_id;
-    msg_temp.header.seq = seq;
+//    msg_temp.header.seq = seq;
```
Comment (ToffeeAlbina-TRI) [inline imu_bno055/src/bno055_i2c_activity.cpp]: ditto

File: imu_bno055/src/bno055_i2c_activity.cpp [hunk 6]
```diff
@@ -217,37 +241,40 @@ bool BNO055I2CActivity::spinOnce() {
...
 
     return true;
 }
 
-bool BNO055I2CActivity::onServiceReset(std_srvs::Trigger::Request &req, std_srvs::Trigger::Response &res) {
+bool BNO055I2CActivity::onServiceReset(std::shared_ptr<std_srvs::srv::Trigger::Request> request,
+                                       std::shared_ptr<std_srvs::srv::Trigger::Response> response) {
     if(!reset()) {
-        ROS_ERROR("chip reset and setup failed");
+        RCLCPP_INFO(get_logger(), "chip reset and setup failed");
```
Comment (ToffeeAlbina-TRI) [inline imu_bno055/src/bno055_i2c_activity.cpp]: should this be rclcpp_error vs info?

File: imu_bno055/src/bno055_i2c_node.cpp [hunk 7]
```diff
@@ -11,55 +11,26 @@
...
-        ros::shutdown();
-        return -3;
-    }
-
     if(!activity->start()) {
-        ROS_ERROR("Failed to start activity");
-        delete nh_priv;
-        delete nh;
-        ros::shutdown();
+        RCLCPP_INFO(activity->get_logger(), "Failed to start activity");
```
Comment (ToffeeAlbina-TRI) [inline imu_bno055/src/bno055_i2c_node.cpp]: error vs info?

File: imu_bno055/launch/imu.launch.py (hunk 1 above)
Comment (clalancette) [reply imu_bno055/launch/imu.launch.py]: Right, we'll want to stick with the original license.  Also, I noticed that upstream changed to the BSD license a few days ago, so we'll probably want to pull that in and stick with BSD as well.

File: imu_bno055/src/bno055_i2c_activity.cpp (hunk 2 above)
Comment (clalancette) [reply imu_bno055/src/bno055_i2c_activity.cpp]: Yeah, just kill off all of the `seq` stuff; we have no plans to bring it back in ROS 2.

File: README.md [hunk 8]
```diff
@@ -33,3 +33,18 @@ The Raspberry Pi hardware I2C does not support clock stretching. You have two op
 
 You may need to add your user to the i2c group, e.g. `sudo usermod -aG i2c nvidia`. It should just work after that.
 
+# ROS2
```
Comment (ToffeeAlbina-TRI) [inline README.md:38]: first part of the instructions should be to check out the ros2 branch.

File: README.md [hunk 9]
```diff
@@ -33,3 +33,18 @@ The Raspberry Pi hardware I2C does not support clock stretching. You have two op
...
+## build with colcon
+`colcon build --event-handler console_direct+ --packages-select  imu_bno055`
+
+## launch
+`ros2 launch imu_bno055 imu.launch.py`
```
Comment (ToffeeAlbina-TRI) [inline README.md]: maybe it's just me, but I'm philosophically opposed to extraneous dots in file names - call it imu_launch.py or IMUlaunch.py (but then caps... yuck)

File: README.md [hunk 10]
```diff
@@ -33,3 +33,18 @@ The Raspberry Pi hardware I2C does not support clock stretching. You have two op
...
+
+## config
+In `config/default_imu_bno055.yaml` you can
```
Comment (ToffeeAlbina-TRI) [inline README.md]: maybe we want to say something like:

A default configuration file listing all settable parameters is provided and is used by the provided default launch file... if you need to change any of the parameters, it is recommended that you create an appropriate configuration file for your system and reference that in an appropriate launch file for your system - the defaults provided here are for reference.

File: README.md (hunk 9 above)
Comment (clalancette) [reply README.md]: There's some disagreement in the community about this: https://discourse.ros.org/t/questioning-the-launch-py-file-extension/6352/8 .

Personally, I'm with you and prefer only a single separator, so I'd also prefer imu_launch.py

File: README.md (hunk 8 above)
Comment (ToffeeAlbina-TRI) [inline README.md:38]: Do we want to say something like:

output topics and service calls provided are the same as for the ROS driver

Comment (ToffeeAlbina-TRI) [reply README.md:38]: Also, up on or before line 3 in this file we should say something like:

for the ROS2 node, follow the instructions in the ROS2 section below

File: imu_bno055/launch/imu.launch.py (hunk 1 above)
Comment (ToffeeAlbina-TRI) [reply imu_bno055/launch/imu.launch.py]: @cuixiongyi you should probably merge in the upstream changes before we move with our internal open source process - we'll need to make sure the lawyers see the actual license that the software is going to be released under.

File: README.md (hunk 9 above)
Comment (cuixiongyi) [reply README.md]: I prefer imu_launch.py as well. I was under under the influence of robot_localization package where they use dot.

Comment (cuixiongyi) [reply README.md]: No, I did not write code under influence.

File: README.md (hunk 8 above)
Comment (cuixiongyi) [reply README.md:38]: I'm wondering should it be a ros2 branch or crystal-devel branch?

Comment (ToffeeAlbina-TRI) [reply README.md:38]: Good question.  I don't think there's a standard yet, but maybe @clalancette knows more?

Comment (cuixiongyi) [reply README.md:38]: fixed.

Comment (ToffeeAlbina-TRI) [reply README.md:38]: robot_localization uses crystal-devel, sophus and ecl_core use /release/1.0-crystal, but those are actually released, so maybe we should go with crystal-devel.

**Review by ToffeeAlbina-TRI — APPROVED** [review]

**Pull request merged**